### DISCUSSION TITLE
Accessing RavenDb Cloud from Azure website

### DISCUSSION THREAD

**Discussion by agfabregat** [ravendb/ravendb#12923] [Q&A]
I have a database in RavenDb Cloud that I access from a web.api C# application in my laptop. 
I've deployed the application to an Azure website.  
I copied the pfx file from my laptop to the Azure website. (it is obviously not that simple)
The application works except when it has to initialize the RavenDb Cloud, it gets a **"Invalid Provider type"** error.
Here is the code, that gets passed the url and the path to the pfx file:

            var documentStore = new Raven.Client.Documents.DocumentStore { Urls = new[] { url } };
            documentStore.Certificate = new X509Certificate2(ravenDbPfxPath);
            documentStore.Database = dbName;
            documentStore.Initialize();                                                //Gets "Invalid Provider type" error

I am totally unfamiliar with all the certificate rigmarole.
Any guidance will be appreciated.

**Comment by haludi** [comment]
Hi,

Can you send us the full error with the stack trace?

Best Regards,
Igal

**Comment by agfabregat** [comment]
Thank you so much for responding, I will be posting the full error later today.  In the meantime, can you tell me if the procedure I followed is the correct one or point me to a document describing how to connect from an Azure website (in an Azure website one does not have full access to the machine).
Than you again
Adolfo

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by haludi** [comment]
Hi,

I am not familiar with this specific infrastructure but in general, your
steps seemed to be ok.
But I really need more information.
Full error with the stack trace will be a good start.

Best regards,
Igal

**Comment by agfabregat** [comment]
Thank you so much for your suggestions.  It turns out there is a process for installing certificates in Azure websites, you go to Configurations for the website and there is an option there for installing a certificate.

Thanks again for your help.
Adolfo


Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by haludi** [comment]
Hi,

Is ravenDbPfxPath a path to the file system?
If so the installation of the certificate is just to copy the file, right?

   - Once again can you send the full error you got with its stack trace? :)

Best regards,
Igal

**Comment by jggutierrez** [comment]
I'm not sure about your specific error, but when creating a certificate on an Azure web app, you have to add an option: `X509KeyStorageFlags.MachineKeySet` . See this discussion: https://groups.google.com/g/ravendb/c/COnMfsxr108

Here's the pattern that I've used successfully:
```c#
public class RavenSettings
  {
      public string[] Urls {get;set;} 
      public string Database {get;set;}
      public string CertificatePath {get;set;}
      public string CertificatePassword {get;set;}
  }

public static class Extensions
{
    public static IServiceCollection AddRaven(this IServiceCollection services, RavenSettings settings)
    {
        var cert = new X509Certificate2(settings.CertificatePath, settings.CertificatePassword, X509KeyStorageFlags.MachineKeySet);
        var docStore = new DocumentStore
        {
            Urls = settings.Urls,
            Database = settings.Database,
            Certificate = cert
        }.Initialize();
        return services.AddSingleton(docStore);
    }

   // read settings from appsettings config section "RavenSettings"
    public static IServiceCollection AddRaven(this IServiceCollection services, IConfiguration config)
    {
        var ravenSettings = config.GetSection("RavenSettings").Get<RavenSettings>();
        return services.AddRaven(ravenSettings);
    }
}
```
